sdk/python/kfp/compiler/v2_compat.py

```diff
@@ -126,13 +126,18 @@ def update_op(op: dsl.ContainerOp,
 
   for artifact_name, spec in sorted(
       component_spec.output_definitions.artifacts.items()):
+    metadata_path = op.file_outputs.get(artifact_name)
+    if not metadata_path:
+        metadata_path = op.output_artifact_paths.get(artifact_name)
+        if not metadata_path:
+            raise Exception(f"Compiler internal error: cannot find output artifact path for artifact '{artifact_name}' in op '{op.name}'")
```

Comment:
The reason I find artifacts in `op.output_artifact_paths` is because `mlpipeline-ui-metadata` artifact has special handling in https://github.com/kubeflow/pipelines/blob/b3c2351a02231f02ecb3f144d3bade06d854ad18/sdk/python/kfp/dsl/_container_op.py#L1206-L1215.

I'm not sure we are ready to remove that handling. Also, when ContainerOps are constructed, we do not know whether it's in v2 compatible mode or v1, therefore the safest fix is to get the artifact here. It seems that this works out at the end, but I'd hope someone knows more about the compiler to double check this.

Reply:
Ohh, thanks for letting me know about `kfp.COMPILING_FOR_V2`. I can try using it instead to remove the hack here.

Reply:
Updated code using `kfp.COMPILING_FOR_V2`, now it's no longer hacky : )